File: 110-balanced-binary-tree/balanced-binary-tree.java
class Solution {
    public boolean isBalanced(TreeNode root) {
        return checkHeight(root) != -1;
    }
    
    private int checkHeight(TreeNode node) {
        if (node == null) {
            return 0;
        }
        
        int leftHeight = checkHeight(node.left);
        if (leftHeight == -1) {
            return -1;
        }
        
        int rightHeight = checkHeight(node.right);
        if (rightHeight == -1) {
            return -1;
        }
        
        // If the height difference is more than 1, return -1 (not balanced)
        if (Math.abs(leftHeight - rightHeight) > 1) {
            return -1;
        }
        
        // Return the height of the current node
        return Math.max(leftHeight, rightHeight) + 1;
    }
}

File: 110-balanced-binary-tree/balanced-binary-tree.cpp
class Solution {
public:
    pair<bool, int> check(TreeNode* root) {
        if (!root) return {true, 0};

        auto left = check(root->left);
        auto right = check(root->right);

        bool balanced = left.first && right.first && abs(left.second - right.second) <= 1;
        int height = 1 + max(left.second, right.second);

        return {balanced, height};
    }

    bool isBalanced(TreeNode* root) {
        return check(root).first;
    }
};
